### PR TITLE
feat: support typescript files with default exports

### PR DESCRIPTION
**Pull request by timcosta**
This PR is the simplest way to introduce TS via ts-node compatibility to haute. It adds the `ts` extension to the list of extensions `require-directory` is allowed to consider, and when the extension is `ts` and there's a `default` prop on the required file it'll use that default prop instead of the required file in order to be compatible with TS.

I attempted to add `.mjs` support as well, however that's going to be more involved and requires deprecating some of the node versions and changing from using `require()` to using `await import()` which likely necessitates a replacement for `require-directory` as well.

I'm a little worried about whether this'll pass CI for earlier node versions. Node 14 can require `.ts` files without exploding, but I'm counting on Travis to check the others.

### CODE REVIEW

**Comment by timcosta** [comment]
Edited: I take this back, CI is working, just takes forever to fire.

**Comment by timcosta** [comment]
Found an issue with `validator` which i suppose i had known about but just forgot about. working on that now

**Comment by devinivy** [comment]
Hey, thanks a lot for thinking through this.  Honestly I am not prepared to make the tradeoff of replacing `require.resolve()` with some custom code in order to support typescript.  Haute is meant to be a very generic utility, and I am hoping not to offer any special typescript support aside from whatever defaults might make life easier for typescript users.  For example, at first it seemed that we might just have to add `'ts'` to the list of default extensions for require-directory and that would do the trick— that would have been an easy win that I'd totally go for.  But some other issues came-up here, i.e. special interpretation of typescript modules to look for default exports, followed by module resolution issues.

That said, I consider it a wart when tooling doesn't interop with pal, and typescript is certainly no exception.  In the spirit of pal we should aim to make haute/haute-couture configurable for typescript users.  I have a proof of concept working with ts-node, haute, haute-couture, and the boilerplate 👍 Here are the changes to haute I'm considering: https://github.com/devinivy/haute/compare/post-process-and-extensions

And given those changes, here's an haute-couture amendment that would bring typescript support to a project:
```js
'use strict'; // .hc.js

const Path = require('path');
const HauteCouture = require('@hapipal/haute-couture');

module.exports = {
    $default: {
        extensions: ['ts', ...HauteCouture.defaults.extensions],
        postProcess: (arg, filename) => {

            if (Path.extname(filename) === '.ts' && arg && 'default' in arg) {
                return arg.default;
            }

            return arg;
        }
    }
};
```

This way we don't need explicit typescript support, but a few lines of code in a project would do the trick.  I'd consider exporting these preset defaults from haute-couture (or possibly even toys?) too if that would be a nice compromise.  Point is: I'd like to handle these situations higher up in the stack, close to the user's app and toolchain rather than in a dependency of a dependency.  Hope this sounds reasonable— looking forward to feedback and figuring this out!

**Comment by timcosta** [comment]
I 100% agree that this change went from being fairly straightforward to being less ideal after the necessary changes, and don't blame you at all for wanting to find a better way to resolve it.

For now we're just forcing module resolution to my branch of haute and the service is very pre-production, so we aren't blocked by this and it's definitely better that we find a maintainable and reasonable change.

There are essentially four issues that exist:
1. `.ts` files that use `export default` which is roughly equivalent is ts-land to `module.exports =` end up exporting an object `{ default: any }` rather than just `any` like requiring a `module.exports` file would return. I believe your change that includes the `postProcess` function would resolve this as well (though the branch seems to use `preProcess` some places and `postProcess` others)
2. `.ts` files in folders arent required by `require-directory` - I believe your change to expose the extensions via HC proposed above would solve this.
3. `.ts` files in the haute-couture root folder (i.e `validator.ts`) use regular `require` instead of `require-directory`, and regular `require` doesnt handle `.ts` files by default in my testing
4. `.hc.ts` isnt read, which is likely the same-ish root cause as the above but in a different module so my haute changes dont resolve this - I just pulled my config into `lib/index.ts` and passed it along with `__dirname` manually. I don't think your proposed change would resolve this either.

tldr of the above block is that I think `.ts` files in the HC root directory will still have problems being required with the above change, and that issue is largely what caused my proposed change to balloon as well.

I'm happy to help contribute towards the fixes for this however, and have a fairly involved project that I can use to test solutions and whatnot in easily. If the determination is made that haute/haute-couture can't/won't be updated to fix this that's totally fine, we'll just go back to the land of _gulp_ manually requiring and registering everything.

**Comment by devinivy** [comment]
> tldr of the above block is that I think `.ts` files in the HC root directory will still have problems being required with the above change, and that issue is largely what caused my proposed change to balloon as well.

In my testing this case worked just fine 👍 I believe ts-node adds to `require.extensions`, so node will treat js and ts files equivalently when it comes to `require()` and `require.resolve()`.  I _think_ the issue in your first few commits was that `normalizeDefaultExport()` was only applied to items that came through `require-directory`.

> `.hc.ts` isnt read

This is a whole other can of worms, ha!  And it goes beyond haute-couture: hpal additionally does look explicitly for `.hc.js`.  I would be open to making it more generic, like a `require.resolve()`ed `.hc`, but again I am not particularly inclined to have special support for typescript in haute-couture and hpal.  Not that it especially matters, but there is a precedent for this in other libraries such as eslint too with `.eslintrc.js`.  Glad you have a workaround! 🙏  Are there downsides to leaving it as `.hc.js` for your purposes?  I know that without this file hpal will not know where your plugin lives and what amendments you're using when utilizing e.g. `hpal make`, and `hpal` is a node script so it is not clear to me how it would pick-up typescript configuration files.  Someone in typescript land may want to make a wrapper around it with a ts-node-script shebang.  Heartening to see that `ts-node --script-mode node_modules/@hapipal/hpal/bin/hpal` seems to do what I would expect it to!

I know you already get where I'm coming from, but I do see these interoperability problems as being mostly on the typescript ecosystem to speak to rather than javascript libraries/authors.  To be clear I absolutely _do_ want typescript users to have ways to configure pal tooling to work for them.  I am totally into doing whatever we can to get out of the way for typescript users and make things in pal land more extensible or generic, as that is one if its design goals, but that's as far as I'm comfortable going for now.  I can tell you're understanding of this, which is appreciated!  I feel sure we'll land on something workable.

**Comment by timcosta** [comment]
> In my testing this case worked just fine 👍

Well this is great to hear, glad it was a me problem haha.

I don't think there are any explicit downsides to leaving `.hc.js` as `js` other than needing to manually hook up types like this:
```js
/// <reference path="./typings/haute-couture/index.d.ts" />
const HC = require('haute-couture');

/** @type {HC.AmendmentConfig} */
module.exports = {
    recursive: true,
    add: [
        {
            place: 'graphql',
            method: 'register',
            list: true,
            recursive: false,
            after: ['plugins'],
            before: ['routes'],
        },
    ],
};

```
The above loads my custom HC typings and then annotates the exported manifest which then has intellisense as one would hope.

**Comment by devinivy** [comment]
Hey @timcosta, just wanted to let you know that this is still on my radar!  I intend to pursue the `postProcess()` path and ship it in haute-couture v4, also backporting to v3 if you're still using that version.  I am still chewing on how to best support `.hc.ts` a little bit, trying to keep in mind that node's ESM support will be more relevant than it is today before we know it.

**Comment by devinivy** [comment]
Thanks for working with me on this @timcosta.  Closing, but we can move over to #18 for further conversation 👍
